pkg/gen/python-templates/yaml.py.mustache

```diff
@@ -64,36 +70,48 @@ class ConfigFile(pulumi.ComponentResource):
         return tables._CASING_BACKWARD_TABLE.get(prop) or prop
 
 
-def read_url(url: str) -> str:
+def _read_url(url: str) -> str:
     response = requests.get(url)
     response.raise_for_status()
 
     return response.text
 
 
-def read_file(path: str) -> str:
+def _read_file(path: str) -> str:
     with open(path, 'r') as file:
         data = file.read()
 
     return data
 
 
-def parse_yaml_document(objects, opts: Optional[pulumi.ResourceOptions] = None):
-    resources = {}
+def _build_resources_dict(objs: List[pulumi.Output]) -> Dict[pulumi.Output, pulumi.Output]:
+    return {key: value for key, value in objs}
+
+
+def _parse_yaml_document(objects, opts: Optional[pulumi.ResourceOptions] = None,
+                         transformations: Optional[List[Callable]] = None) -> pulumi.Output:
+    objs = []
     for obj in objects:
-        # TODO: transformation support
-        file_objects = parse_yaml_object(obj, opts)
-        for key, value in file_objects:
-            resources[key] = value
+        file_objects = _parse_yaml_object(obj, opts, transformations)
+        for file_object in file_objects:
+            objs.append(file_object)
+
+    return pulumi.Output.all(*objs).apply(_build_resources_dict)
 
-    return resources
 
+def _parse_yaml_object(obj, opts: Optional[pulumi.ResourceOptions] = None,
+                       transformations: Optional[List[Callable]] = None) -> [pulumi.Output]:
+    """
+    _parse_yaml_object parses a YAML manifest object, and creates the specified resources.
+    """
 
-def parse_yaml_object(obj, opts: Optional[pulumi.ResourceOptions] = None):
     if not obj:
         return []
 
-    # TODO: transformation support
+    # Allow users to change API objects before any validation.
+    if transformations is not None:
+        for t in transformations:
+            obj = t(obj)
```

Comment:
Reading this code now, I realize that we're running the transformations _before_ we unpack the lists of resources, both here and in the nodejs SDK. I think we want to do the opposite? Perhaps not in this PR but in the future.